Add rollaxis and moveaxis

TAdeJong opened the pull request:
Attempts to fix #2559 by aliasing the numpy functions, which depend on the transpose method.
Although #2955 existed as a WIP, I moved forward with a new PR with the easier aliasing method.

- [x] Tests added / passed
- [x] Passes `flake8 dask`

TAdeJong commented:
Travis fails on Python3.7 only with a recursion error that I do not entirely understand or reproduce locally...

Can the `assert isinstance()` in combination with the `getattr` be a problem?

TAdeJong commented:
Ah, that makes sense now and then a straight up alias `moveaxis=np.moveaxis` would make sense as @mrocklin suggested. 
One thing I am not sure about: would just plainly in the place where the `def` is now be the right place to do such an alias? I will try to adapt this tomorrow. (Getting late here)

jakirkham commented:
Thanks for picking this up, @TAdeJong. Agree that trying a simple alias is a good idea. Though we may still run into issues with `__array_function__`. That said, it is something easy to try and if it doesn't work we can give it more thought.

TAdeJong commented:
OK, so a simple alias at least prevents infinite recursion, but in the case of `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION='1'`, we have a
`TypeError: no implementation found for 'numpy.moveaxis' on types that implement __array_function__: [<class 'dask.array.core.Array'>]`
Seems the __array_function__ actually makes things a bit worse. I have not wrapped my head around NEP 18 entirely yet, anyone any suggestions?

jakirkham commented:
Thanks for trying this out @TAdeJong. Yeah this is what I was alluding to above. Will take a look and see what options we have to move forward. Though others should feel free to jump in if they have thoughts as well.

pentschev commented:
The `__array_function__` dispatches NumPy functions to other APIs based on the types of the arrays. In the case here, the array is a Dask array, so it will dispatch to Dask. Since you were wrapping around a NumPy function, NumPy will dispatch to Dask, which will in turn call the NumPy function, and you get a recursion. If you only alias it, then there's no Dask implementation, as you've mentioned in your last comment.

I think the only solution for now would be to dispatch work through some mechanism such as `da.map_blocks`. I'm not too familiar with such Dask mechanisms, so maybe that isn't the most appropriate and there may be a better one for this particular case.

jakirkham commented:
Thanks @pentschev. Your description of the problem sounds correct. Though we might have a better option for implementation.

After taking a quick look, it seems all NumPy functions as of NEP 18 have a `__skip_array_function__ ` ( https://github.com/numpy/numpy/pull/13389 ) method. I think if we try to alias that it should work to bypass NEP 18. However we will need to handle the case that an older version of NumPy is in use by falling back to the current aliases in this PR if `__skip_array_function__` is not defined.

Let's check with @shoyer to see if this is a reasonable path forward or if there is something else we should be doing.

pentschev commented:
`__skip_array_function__` is still open https://github.com/numpy/numpy/pull/13585, that means it won't be available at least before NumPy 1.17, which will prevent this PR from getting in until then.

My proposal was to find a solution that would work independent of that. I think `__skip_array_function__` is the best solution for this situation, unfortunately not usable yet.

TAdeJong commented:
So I guess we do _not_ need to cover for the case `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION='1'` and `__skip_array_function__` undefined? In that case it seems a reasonable `try...except`. 

I am fine with leaving this open until we have `__skip_array_function__` . 
To avoid confusion it _is_ probably a good idea to close #2955 though.

pentschev commented:
IMO, the test could be skipped for now to get this PR merged. This can be done by checking if NEP-18 is active, just as in https://github.com/dask/dask/blob/master/dask/array/tests/test_array_function.py#L5-L11

jakirkham commented:
> `__skip_array_function__` is still open [numpy/numpy#13585](https://github.com/numpy/numpy/pull/13585), that means it won't be available at least before NumPy 1.17, which will prevent this PR from getting in until then.

That seems like a fair assessment Peter.

> My proposal was to find a solution that would work independent of that. I think `__skip_array_function__` is the best solution for this situation, unfortunately not usable yet.

Sure. That would amount to resuming PR ( https://github.com/dask/dask/pull/2955 ) and adding a similar PR for `moveaxis`. I'm not sure it is worth the effort if we are going to remove it soon anyways. Though wouldn't be opposed to doing it if we decide it is useful.

> So I guess we do _not_ need to cover for the case `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION='1'` and `__skip_array_function__` undefined? In that case it seems a reasonable `try...except`.

IIRC this will become default behavior in NumPy 1.17. So assuming `__skip_array_function__` lands in that release. We can just check for that.

> I am fine with leaving this open until we have `__skip_array_function__` .

Sounds good.

> To avoid confusion it _is_ probably a good idea to close #2955 though.

I guess that depends on whether we need a stopgap as Peter alluded to before. Do you have thoughts on this?

jakirkham commented:
> IMO, the test could be skipped for now to get this PR merged. This can be done by checking if NEP-18 is active, just as in

I'm not sure I completely agree unfortunately. It seems this function will break on different versions of NumPy and we don't have a good way to explain to the user why. Would prefer to see a solution that smoothed over these differences.

pentschev commented:
> I'm not sure I completely agree unfortunately. It seems this function will break on different versions of NumPy and we don't have a good way to explain to the user why. Would prefer to see a solution that smoothed over these differences.

But having NEP-18 is an opt-in only condition, and the user needs to explicitly set `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1`, which is experimental, so they should be aware of issues. However, I don't have a strong opinion, if everybody is fine with keeping this open, that's fine by me too, otherwise if someone is in need of the functionality provided in this PR, we have an alternative to allow things to move forward with a bit of compromise.

TAdeJong commented:
In terms of _needing_ these functions in particular, I suspect the need is fairly limited as in most cases a similar `.transpose()` function is a possibility. Although I am actually somewhat surprised that these functions are the first where we walk into this NEP-18 issue.

pentschev commented:
It's the first time in Dask, but we ran into that in CuPy https://github.com/cupy/cupy/issues/2029, which was actually the trigger to have `__skip_array_function__`.

jakirkham commented:
NEP 18 is still really new. So am not surprised we are running into a few speed bumps. We are all still learning what this means and what it should look like (hence why it is experimental ;). Hopefully we will have this sort of thing worked out by NumPy 1.17 when the protocol most likely will be enabled by default.

jakirkham commented:
> ...if someone is in need of the functionality provided in this PR, we have an alternative to allow things to move forward with a bit of compromise.

I think if someone needs these, we can just say the NumPy functions already work (the gist of this PR). They should try using them as is on Dask Arrays. If that doesn't work, we should try to get a bit more information about why not.

pentschev commented:
> I think if someone needs these, we can just say the NumPy functions already work (the gist of this PR).

That's a good point. As I mentioned before, I've got no objections in waiting on this PR nor skipping the tests. If we can wait, let's wait.

martindurant commented:
Should this be closed, then, for now? I don't follow quite what the plan is.

pentschev commented on dask/array/routines.py:
```diff
@@ -290,6 +290,9 @@ def matmul(a, b):
     return out
 
 
+moveaxis = np.moveaxis
```
```suggestion
def numpy_implementation(func, *args, **kwargs):
    if hasattr(func, '_implementation'):
        return func._implementation(*args, **kwargs)
    else:
        return func(*args, **kwargs)

moveaxis = partial(numpy_implementation, np.moveaxis)
```

pentschev replied:
Probably `numpy_implementation` could go somewhere more general, such as `dask/utils.py`.

jakirkham replied:
Seems like a good idea. Could add to [dask/array/numpy_compat.py]( https://github.com/dask/dask/blob/master/dask/array/numpy_compat.py ).

shoyer replied:
Please don't use `_implementation`. This is not a public API exposed by NumPy, as indicated by the name starting with an underscore. (If this isn't clear from the NEP, we should fix that.)

It would be more maintainable to copy the implementations from NumPy. These are not complicated functions, so I really don't think there is much that could go wrong with copying their code.

pentschev replied:
Is there an alternative to copying the implementation? These are not the only cases where this will be necessary. In CuPy as well we need this for some functions that are currently aliases, I've submitted a PR for those https://github.com/cupy/cupy/pull/2249.

Granted, I was very inattentive to `_implementation` being non-public.

pentschev commented on dask/array/routines.py:
```diff
@@ -971,6 +974,9 @@ def roll(array, shift, axis=None):
     return result
 
 
+rollaxis = np.rollaxis
```
```suggestion
rollaxis = partial(numpy_implementation, np.rollaxis)
```

pentschev commented:
I think we shouldn't close this issue, but maybe add `[WIP]` to the title, just so it becomes visible it can't be merged yet?

NumPy master has now an `_implementation` attribute for functions, that allows falling back to NumPy even when `__array_function__` is enabled. I've added suggestions to replace the aliasing of `moveaxis` and `rollaxis` by a dispatcher for NumPy's `_implementation`.

Once NumPy 1.17 is merged, this change should automatically work there. However, it's important to note that these two functions will never work for NumPy 1.16.x when `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1`, but probably when NumPy 1.17 is released, we should use that for the failing test, since it will have better `__array_function__` support, which is what we are aiming for with Dask too.

shoyer commented:
Currently there is no alternative to copying the implementation (and there
won't be, at least for NumPy 1.17). I really wanted to make this work, but
it presented many design complications for NumPy...

On Thu, Jun 20, 2019 at 11:49 PM Peter Andreas Entschev <
notifications@github.com> wrote:

> *@pentschev* commented on this pull request.
> ------------------------------
>
> In dask/array/routines.py
> <https://github.com/dask/dask/pull/4822#discussion_r295994103>:
>
> > @@ -290,6 +290,9 @@ def matmul(a, b):
>      return out
>
>
> +moveaxis = np.moveaxis
>
> Is there an alternative to copying the implementation? These are not the
> only cases where this will be necessary. In CuPy as well we need this for
> some functions that are currently aliases, I've submitted a PR for those
> cupy/cupy#2249 <https://github.com/cupy/cupy/pull/2249>.
>
> Granted, I was very inattentive to _implementation being non-public.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/4822?email_source=notifications&email_token=AAJJFVTLKTUASEXOJ2LNPWTP3PUOXA5CNFSM4HN47PT2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4GVKWI#discussion_r295994103>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJJFVT5GE24GNXKJGXNG23P3PUOXANCNFSM4HN47PTQ>
> .
>

shoyer commented:
For something like the cupy functions, you should be able to make a small
wrapper function that calls the NumPy functions on dtype attributes of CuPy
arrays.

On Thu, Jun 20, 2019 at 11:53 PM Stephan Hoyer <shoyer@gmail.com> wrote:

> Currently there is no alternative to copying the implementation (and there
> won't be, at least for NumPy 1.17). I really wanted to make this work, but
> it presented many design complications for NumPy...
>
> On Thu, Jun 20, 2019 at 11:49 PM Peter Andreas Entschev <
> notifications@github.com> wrote:
>
>> *@pentschev* commented on this pull request.
>> ------------------------------
>>
>> In dask/array/routines.py
>> <https://github.com/dask/dask/pull/4822#discussion_r295994103>:
>>
>> > @@ -290,6 +290,9 @@ def matmul(a, b):
>>      return out
>>
>>
>> +moveaxis = np.moveaxis
>>
>> Is there an alternative to copying the implementation? These are not the
>> only cases where this will be necessary. In CuPy as well we need this for
>> some functions that are currently aliases, I've submitted a PR for those
>> cupy/cupy#2249 <https://github.com/cupy/cupy/pull/2249>.
>>
>> Granted, I was very inattentive to _implementation being non-public.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/dask/dask/pull/4822?email_source=notifications&email_token=AAJJFVTLKTUASEXOJ2LNPWTP3PUOXA5CNFSM4HN47PT2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4GVKWI#discussion_r295994103>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAJJFVT5GE24GNXKJGXNG23P3PUOXANCNFSM4HN47PTQ>
>> .
>>
>

pentschev commented:
@shoyer I totally missed most of the discussion about that, but I'm assuming `_implementation` is non-public due to some issue with its use. If it's too complicated to explain briefly, would you mind pointing me to discussion about its consequences and downsides?

shoyer commented:
See this summary from the mailing list:
https://mail.python.org/pipermail/numpy-discussion/2019-May/079541.html
(which is also linked to from the NEP, in the section "Partial
implementation of NumPy's API")

On Thu, Jun 20, 2019 at 11:58 PM Peter Andreas Entschev <
notifications@github.com> wrote:

> @shoyer <https://github.com/shoyer> I totally missed most of the
> discussion about that, but I'm assuming _implementation is non-public due
> to some issue with its use. If it's too complicated to explain briefly,
> would you mind pointing me to discussion about its consequences and
> downsides?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/4822?email_source=notifications&email_token=AAJJFVRQRMQXZBTE2KVV3NTP3PVQZA5CNFSM4HN47PT2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYGS6JA#issuecomment-504180516>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJJFVQDTQULEFBJYDNN4JLP3PVQZANCNFSM4HN47PTQ>
> .
>

jakirkham commented:
Thanks for the feedback, Stephan. In that case, @TAdeJong would you be ok copying the NumPy implementations and putting them both in [dask/array/numpy_compat.py]( https://github.com/dask/dask/blob/master/dask/array/numpy_compat.py )?

TAdeJong commented:
I will use some time next week for that.

pentschev commented:
Thanks for the link @shoyer, that was a good summary. It's really unfortunate that we can't do that for NumPy 1.17, but thanks for all the work you did on that!

pentschev commented:
@jakirkham @TAdeJong if we copy the implementation, we then will have no dependencies on NumPy 1.17, so we can merge this once that is done.

TomAugspurger commented:
Where does this PR stand now? We had a user report a warning (I think from NumPy) that move axis wasn't implemented.

TAdeJong commented:
I hope to spend some time this weekend to make this mergable.

TAdeJong commented:
I now added both implementations from numpy. I added them to array/routines. @jakirkham, was there any specific reason you had thought to put them in numpy_compat?

jakirkham commented:
Two reasons, we copy other functions there and it contains the proper NumPy license.

TomAugspurger commented:
Can we also add a test ensuring that `np.rollaxis(dask.array.Array)` is lazy and returns a Dask Array?

```python
def test_move_axis_numpy():
    a = da.random.random((4, 4, 4), chunks=2)
    result = np.moveaxis(a, 2, 0)
    assert isinstance(result, da.Array)
    assert_eq(result, np.moveaxis(a.compute(), 2, 0))
```

I can also push that here if you want.

pentschev commented:
LGTM, I would recommend merging it.

Thanks for the work @TAdeJong, and @TomAugspurger for the latest adjustments!

TomAugspurger commented:
Thanks @TAdeJong!
